Add `.get(range)`

Philippe-Cholet opened the pull request:
Closes #447.

So much changes happened since this work was initiated that rebase on top of master seems nothing less than a nightmare to me (I know, I tried). I therefore looked at the git differences, committed those (after formatting) and credited @wod3 and @TrolledWoods for their respective commits. Then I fixed my own review and a bit more.

---

I thought I would additionally implement `IteratorIndex` for `Either<range, range>` (and `Box<dyn range>` but I could not) but doing so would later result in a nasty breaking change if this was removed after an eventual inclusion to `Iterator`.

<details>
<summary>Details about either</summary>

```rust
    impl<L: Sealed, R: Sealed> Sealed for either::Either<L, R> {}


impl<I, L, R> IteratorIndex<I> for Either<L, R>
where
    I: Iterator,
    L: IteratorIndex<I>,
    R: IteratorIndex<I>,
{
    type Output = Either<<L as IteratorIndex<I>>::Output, <R as IteratorIndex<I>>::Output>;

    fn index(self, iter: I) -> Self::Output {
        // Could be more succinct with `map_either_with` but it would require to bump `either` version.
        match self {
            Either::Left(left) => Either::Left(left.index(iter)),
            Either::Right(right) => Either::Right(right.index(iter)),
        }
    }
}


    /// // It works nicely with either.
    /// use itertools::Either;
    ///
    /// let mut either_range = Either::Left(0..1);
    /// range = vec.iter().get(either_range).copied().collect();
    /// assert_eq!(&range, &[3]);
    ///
    /// either_range = Either::Right(2..);
    /// range = vec.iter().get(either_range).copied().collect();
    /// assert_eq!(&range, &[4, 1, 5]);
```
</details>

Philippe-Cholet commented on src/iter_index.rs:
```diff
@@ -0,0 +1,114 @@
+    type Output = Take<Skip<I>>;
+
+    fn index(self, iter: I) -> Self::Output {
+        debug_assert!(!self.is_empty(), "The given `RangeInclusive` is exhausted. The result of indexing with an exhausted `RangeInclusive` is unspecified.");
```
**Feedback from CI:** `RangeInclusive::is_empty` requires Rust 1.47.0+ so the MSRV check fails.
It can't be inlined because of a private field (`exhausted`). Our options:
- Wait that the MSRV hits 1.47.0 to merge this.
- (Merge this feature without `RangeInclusive`, it would be really weird for the user).
- Or temporarily disable this `debug_assert`:
```suggestion
        // TODO: Once MSRV >= 1.47.0, uncomment:
        // debug_assert!(!self.is_empty(), "The given `RangeInclusive` is exhausted. The result of indexing with an exhausted `RangeInclusive` is unspecified.");
```

Philippe-Cholet replied:
@jswrenn What do you think we should do here?

jswrenn replied:
So long as we document that the result of indexing with an exhausted `RangeInclusive` is unspecified, we can remove the `debug_assert!`.

Philippe-Cholet replied:
The documentation of `Itertools::get` mentions it so I'm gonna remove this.

Philippe-Cholet commented on src/iter_index.rs:
```diff
@@ -0,0 +1,113 @@
+    type Output = Take<Skip<I>>;
+
+    fn index(self, iter: I) -> Self::Output {
+        iter.skip(self.start)
+            .take(self.end.saturating_sub(self.start))
```
I thought so too but it directly came from #447. I thought there was a reason for it, and that it would be written in the discussion, but after skimming it, it does not seem to be the case, weird.

Philippe-Cholet commented on src/lib.rs:
```diff
@@ -502,6 +505,53 @@ pub trait Itertools: Iterator {
+    ///
+    /// It's a generalisation of [`Iterator::take`] and [`Iterator::skip`],
+    /// and uses these under the hood.
+    /// Therefore, the resulting iterator is [`DoubleEndedIterator`]
+    /// and/or [`ExactSizeIterator`] if the adapted iterator is.
```
I'm pretty sure I verified this but it definitely deserves a written test.

Philippe-Cholet commented:
I rebased on master and added 3 commits: 
- Is there a reason to prefer `Take<Skip>` over `Skip<Take>`? As `Skip<Take>` is simpler defined!
- I don't see how we should really handle a range that includes `usize::MAX` other than with **panic**.
- Clarify when the resulting iterator is `ExactSizeIterator` and/or `DoubleEndedIterator`.

jswrenn commented on src/iter_index.rs:
```diff
@@ -0,0 +1,113 @@
+/// Returns an iterator over a subsection of the iterator.
+///
+/// See [`Itertools::get`] for more information.
+pub fn get<I, R>(iter: I, index: R) -> R::Output
```
I think the name is appropriate — it 'gets' elements from the iterator. We use `get` for consistency with `[T]::get`, and match its API as closely as possible. There's little risk of conflict, since slices do not directly implement `Iterator`.

Philippe-Cholet replied:
Ah fair enough, we don't really need to make the free function `get` public.
